application/ledger_query_service/src/ethereum/block_processor.rs

```diff
@@ -0,0 +1,183 @@
+use crate::{
+    ethereum::queries::{
+        EthereumBlockQuery, EthereumTransactionLogQuery, EthereumTransactionQuery,
+    },
+    web3::{
+        self,
+        types::{Block, Transaction},
+    },
+    ArcQueryRepository, QueryMatch, QueryMatchResult,
+};
+use ethereum_support::web3::{transports::Http, Web3};
+use futures::{
+    future::Future,
+    stream::{self, Stream},
+};
+use std::sync::Arc;
+use tokio;
+
+type BlockQueryResults = Vec<QueryMatch>;
+type TransactionQueryResults = Vec<QueryMatch>;
+type TransactionLogQueryResults = Vec<QueryMatch>;
+
+pub fn process(
+    block_queries: ArcQueryRepository<EthereumBlockQuery>,
+    transaction_log_queries: ArcQueryRepository<EthereumTransactionLogQuery>,
+    transaction_queries: ArcQueryRepository<EthereumTransactionQuery>,
+    client: Arc<Web3<Http>>,
+    block: &Block<Transaction>,
+) -> Result<
+    (
+        BlockQueryResults,
+        TransactionQueryResults,
+        TransactionLogQueryResults,
+    ),
+    (),
+> {
+    let block_query_results = process_block_queries(block_queries, block);
+
+    let transaction_queries = transaction_queries.clone();
+    let transaction_query_results: TransactionQueryResults = block
+        .transactions
+        .iter()
+        .map(move |transaction| {
+            process_transaction_queries(transaction_queries.clone(), transaction)
+        })
+        .flatten()
+        .collect();
+    let result = process_transaction_log_queries(transaction_log_queries, client, block);
+
+    let transaction_log_query_results = match result {
+        Ok(result) => result,
+        Err(e) => {
+            error!(
+                "Could not execute transaction log queries on block {:?}: {:?}",
+                block, e
+            );
+            vec![]
+        }
+    };
+
+    Ok((
+        block_query_results,
+        transaction_query_results,
+        transaction_log_query_results,
+    ))
+}
+
+fn process_block_queries(
```

Comment:
:wrench: Since it is not a trait, it might be worth to return `impl Iterator` and don't `collect` because you are gonna iterate over it again anyway so you can save some performance and intermediary allocations here.

Reply:
Applies to all three functions.